script/_ECS_WIP_TEMP/Shared/Rendering/Material.lua

```diff
@@ -0,0 +1,82 @@
+local AutoShaderVar = require("_ECS_WIP_TEMP.Shared.Rendering.AutoShaderVar") --!temp path
+local Texture = require("_ECS_WIP_TEMP.Shared.Rendering.Texture") --!temp path
+
+---@class Material
+---@field materialName string -- Might use for Specific Material Cache?
+---@field vertexName string -- 'res/shader/vertex/'
+---@field fragmentName string -- 'res/shader/fragment/'
+---@field textures table<Texture>
+---@field shaderState ShaderState
+---@field autoShaderVars table<AutoShaderVar>
+
+---@param self Material
+---@param materialName string
+---@param vertexName string
+---@param fragmentName string
+---@class Material
+---@overload fun(self: Material, materialName: string, vertexName: string, fragmentName: string): Material class internal
+---@overload fun(materialName: string, vertexName: string, fragmentName: string): Material class external
+local Material = Class(function(self, materialName, vertexName, fragmentName)
+    self.materialName = materialName
+    self.vertexName = vertexName
+    self.fragmentName = fragmentName
+    self.textures = {}
+    self.autoShaderVars = {}
+    local shader = Cache.Shader(vertexName, fragmentName)
+    self.shaderState = ShaderState.Create(shader)
+end)
+
+---@param uniformName string
+---@param renderFn function
+---@return AutoShaderVar|nil
+function Material:addAutoShaderVar(uniformName, renderFn)
+    if self.shaderState:shader():hasVariable(uniformName) then
+        local autoShaderVar = AutoShaderVar(uniformName, renderFn)
+        autoShaderVar:setUniformInt(self.shaderState:shader():getVariable(uniformName))
```

Comment:
@LukeJHarvey I would pass the uniform index to the AutoShaderVar constructor instead, then just remove `setUniformInt`. We don't ever want to change the uniform index once we create this AutoShaderVar